### PR TITLE
IP: allow connections with src port == dst port

### PR DESCRIPTION
**Pull request by dgloeck**
The rewrite of is_invalid_packet() in commit 1be2706a93f1 ("net/connection: Modify input function") changed the behavior for packets where the source and destination IP addresses differ but the source and destination ports are the same. The correct behavior is of course to accept these packets. An example where this is required is the zperf sample, which by default binds and connects to port 5001.

As those packets with source and destination IP address and port equal are ruled out early, the default return value can be changed to false. This also allows to drop the code for the AF_PACKET and AF_CAN packets, that returned false for all packets.

### CODE REVIEW

**Review by jukkar — COMMENTED** [review]
It is quite difficult to understand this function now, so I looked the original version in commit 68ef8f06a5082ee3baa4128c4fe8e84832f22a57. Could we try to mimic what that commit did here?

**Comment by dgloeck** [comment]
Actually I think the current function is quite easy to understand:
 - all packets where source and destination port are different are accepted
 - then there is a per-protocol check that detects invalid packets (currently either src == dst or received from self)
 - if no protocol rejected the packet, it is accepted (when this PR is merged)

The function was never meant to detect all invalid packets. Maybe another name would have been better.

**Comment by marc-hb** [comment]
Here are a couple reasons why this function is harder to read than it could be:
- `is_invalid() { return false; ...; return false; ... }` are a double negatives
- `_addr_cmp()` works opposite of `strcmp(), memcmp(),` etc!
- as you wrote, it's not clear whether it's meant to catch all issues or not.
- zero comment; none mentioning the original purpose (resource exhaustion)

> looked the original version in commit 68ef8f0...

... which was at least shorter.

> Maybe another name would have been better.

`is_packet_sane()` ? git grep finds a single call in the entire codebase and I doubt this is part of some API.

**Review by tbursztyka — APPROVED** [review]

**Comment by jukkar** [comment]
I think @marc-hb summarized it well. At least for me, the original version is much easier to understand than this new one, especially as there are no comments why we are doing the checks. In the original version the variable names and the commit message made the function understandable. But as the function has been reworked a lot, it is not so clear to see the idea.
So, could you refactor this function more? I know this is not your code originally, but as you now touched it, we could try to make it more clearer.

**Review by jukkar — CHANGES_REQUESTED** [review]
I think some refactoring is needed here for the `is_invalid_packet()` function.

**Comment by tbursztyka** [comment]
the 3 first tests could be reworked, so:
`if (src_port == dst_port)  {
/* add the 2 other tests there*/
}`

Maybe more readable.

Renaming to is_packet_sane is not a better name. (is_packet_invalid isn't good either).

What it tests here is if src_port and dst_port are same: then ip src/dst should not be same. That's all.

**Comment by marc-hb** [comment]
> Renaming to is_packet_sane is not a better name

Do you mean you don't see any difference between the English words "valid" and "sane"?

For me "[in]valid" implies exhaustive checks and a final decision (as pointed out by @dgloeck), whereas "sane" does not, it refers only to worst issues and matches the very first addition of this function.

**Comment by jukkar** [comment]
I have an alternate version for this at #14737, please check and comment.

**Comment by pfalcon** [comment]
I believe this issue is fixed in alternative way by https://github.com/zephyrproject-rtos/zephyr/pull/14737 . While it's not merge as of now, it's approved and pending merging, whereas this patch is -1'ed.

Please let me close this, as otherwise it's hard to keep track of what PRs are active and which aren't, in these hot last weeks before 1.14. If there's a mistake or misunderstanding, please reopen.

**Pull request closed without merging**
